p2p/client.go

```diff
@@ -625,42 +635,68 @@ func (c *Client) retrieveBlockSyncLoop(ctx context.Context, msgHandler BlockSync
 					c.logger.Error("Blocksync LoadBlock", "err", err)
 					continue
 				}
-
 				c.logger.Debug("Blocksync block received ", "height", h)
-				msgHandler(&block)
+
 				state, err := c.store.LoadState()
 				if err != nil {
 					return
 				}
+				if err := block.Validate(state.GetProposerPubKey()); err != nil {
+					c.logger.Error("Failed to validate blocksync block.", "height", block.Block.Header.Height)
+					continue
+				}
+				msgHandler(&block)
 				h = max(h, state.NextHeight()-1)
 			}
 			c.blocksReceived.RemoveBlocksReceivedUpToHeight(state.NextHeight())
+
 		}
 	}
 }
 
 // advertiseBlockSyncCids is used to advertise all the block identifiers (cids) stored in the local store to the DHT on startup
 func (c *Client) advertiseBlockSyncCids(ctx context.Context) {
-	state, err := c.store.LoadState()
-	if err != nil {
-		return
-	}
-	baseHeight, err := c.store.LoadBaseHeight()
-	if err != nil {
-		return
-	}
-	for h := baseHeight; h <= state.Height(); h++ {
+	ticker := time.NewTicker(c.conf.BlockSyncRequestIntervalTime)
+	defer ticker.Stop()
 
-		id, err := c.store.LoadBlockCid(h)
-		if err != nil || id == cid.Undef {
-			continue
-		}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			// if no connected at p2p level, it will try again after ticker time
+			if len(c.Peers()) == 0 {
+				continue
+			}
+			state, err := c.store.LoadState()
+			if err != nil {
+				return
+			}
+			baseHeight, err := c.store.LoadBlockSyncBaseHeight()
+			if err != nil && !errors.Is(err, gerrc.ErrNotFound) {
+				return
+			}
+			for h := baseHeight; h <= state.Height(); h++ {
+				if ctx.Err() != nil {
+					return
+				}
+				id, err := c.store.LoadBlockCid(h)
+				if err != nil || id == cid.Undef {
+					continue
+				}
+				revision := uint64(0)
+				if h >= state.RevisionStartHeight {
```

Comment:
it's wrong for multiple revisions
this assumes `revision` is either `0` or the current state revision

Reply:
true, forgot to fix this. i changed it so it loads the block and advertises according block revision.